components/function-controller/internal/git/go2git.go

```diff
@@ -0,0 +1,152 @@
+package git
+
+import "C"
+import (
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+
+	git2go "github.com/libgit2/git2go/v31"
+	"github.com/pkg/errors"
+)
+
+const (
+	tempDir          = "/tmp"
+	branchRefPattern = "refs/remotes/origin"
+)
+
+type Options struct {
+	URL       string
+	Reference string
+	Auth      *AuthOptions
+}
+
+type cloner interface {
+	cloneRepo(options Options, outputPath string) (*git2go.Repository, error)
+}
+
+type Git2GoClient struct {
+	cloner
+}
+
+func NewGit2Go() *Git2GoClient {
+	return &Git2GoClient{
+		cloner: &git2goCloner{},
+	}
+}
+
+func (g *Git2GoClient) LastCommit(options Options) (string, error) {
+	//commit
+	_, err := git2go.NewOid(options.Reference)
+	if err == nil {
+		return options.Reference, nil
+	}
+
+	tmpPath, err := ioutil.TempDir(tempDir, "fn-git")
+	if err != nil {
+		return "", err
+	}
+	defer removeDir(tmpPath)
+
+	repo, err := g.cloner.cloneRepo(options, tmpPath)
+	if err != nil {
+		return "", errors.Wrap(err, "while cloning the repository")
+	}
+
+	//branch
+	ref, err := g.lookupBranch(repo, options.Reference)
+	if err == nil {
+		return ref.Target().String(), nil
+	}
+	if !git2go.IsErrorCode(err, git2go.ErrNotFound) {
+		return "", err
+	}
+
+	//tag
+	ref, err = repo.References.Dwim(options.Reference)
+	if err == nil {
+		return ref.Target().String(), nil
+	}
+	if !git2go.IsErrorCode(err, git2go.ErrNotFound) {
+		return "", errors.Wrap(err, "while lookup branch")
```

Comment:
While lookup branch? :P Not tag? And maybe we should wrap other errors in the whole method